pybandits/simulator.py

```diff
@@ -0,0 +1,592 @@
+# MIT License
+#
+# Copyright (c) 2022 Playtika Ltd.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+import os.path
+import random
+from abc import ABC, abstractmethod
+from functools import cached_property
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from bokeh.core.enums import Palette
+from bokeh.layouts import column, row
+from bokeh.models import ColumnDataSource, HoverTool, Legend, Plot, TabPanel
+from bokeh.palettes import Category10, Turbo256
+from bokeh.plotting import figure
+from loguru import logger
+
+from pybandits.base import ActionId, BinaryReward, PyBanditsBaseModel
+from pybandits.mab import BaseMab
+from pybandits.pydantic_version_compatibility import (
+    PYDANTIC_VERSION_1,
+    PYDANTIC_VERSION_2,
+    NonNegativeInt,
+    PositiveInt,
+    PrivateAttr,
+    field_validator,
+    model_validator,
+    pydantic_version,
+)
+from pybandits.utils import visualize_via_bokeh
+
+
+class Simulator(PyBanditsBaseModel, ABC):
```

Comment:
I suggest to split the class methods with `#######` to differentiate input validators VS class/object methods as you did for `mab.py`